partcad/src/partcad/actions/part_actions.py

```
@@ -140,3 +145,160 @@
+        else:
+            source_path = old_path
+
+    # Copy the file into the target package.
+    target_path = (Path(project.path) / new_file_name).resolve()
+    if target_path.exists() and source_path.samefile(target_path):
+        pc_logging.warning(f"Skipping copy: source and target are the same ({source_path}).")
+    else:
+        try:
+            shutil.copy2(source_path, target_path)
+        except shutil.Error as e:
+            raise ValueError(f"Failed to copy '{source_path}' → '{target_path}': {e}")
```

Ruff (0.8.2)

Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling

(B904)